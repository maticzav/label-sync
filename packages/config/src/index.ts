export { LSCConfiguration, LSCRepositoryConfiguration, LSCRepository, LSCLabel } from './types'
export {
  parseConfig,
  isConfigRepo,
  getPhysicalRepositories,
  normalizeColor,
  getLSConfigRepoName,
  LS_CONFIG_PATH,
} from './parse'
