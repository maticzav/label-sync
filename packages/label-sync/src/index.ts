export { repo, label } from './generator'
export { labelsync } from './make'
export * from './presets'
