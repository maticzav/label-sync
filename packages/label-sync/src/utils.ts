export type Dict<T> = { [key: string]: T }
export type Maybe<T> = T | null

/**
 * Creates a fallback default value.
 * @param fallback
 * @param value
 */
export function withDefault<T>(fallback: T, value: T | undefined): T {
  if (value) return value
  else return fallback
}

/**
 * Maps entries in a map.
 * @param m
 * @param fn
 */
export function mapEntries<K, T, V>(
  m: Map<K, T>,
  fn: (v: T, key: K) => V,
): Map<K, V> {
  const map = new Map<K, V>()

  for (const key of m.keys()) {
    map.set(key, fn(m.get(key)!, key))
  }

  return map
}
