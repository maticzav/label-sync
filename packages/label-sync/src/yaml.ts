import yaml from 'js-yaml'
import { Configurable } from './configurable'

export abstract class YAML<T> extends Configurable<T> {
  getYAML(): string {
    const cleanJSON = JSON.parse(JSON.stringify(this.getConfiguration()))
    return yaml.dump(cleanJSON)
  }
}
