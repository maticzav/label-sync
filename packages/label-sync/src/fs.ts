import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import { Maybe } from './utils'

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export function findFolderUp(dir: string, name: string): Maybe<string> {
  return findUp(dir, (p) => fs.lstatSync(p).isDirectory() && path.basename(p) === name)
}

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export function findFileUp(dir: string, name: string): Maybe<string> {
  return findUp(dir, (p) => fs.lstatSync(p).isFile() && path.basename(p) === name)
}

/**
 * Traverses file system up to find a specific directory that contains
 * a matching pattern.
 *
 * @param dir
 * @param pattern
 */
export function findUp(dir: string, pattern: (path: string) => boolean): Maybe<string> {
  switch (path.normalize(dir)) {
    /* End case: we reached the root. */
    /* istanbul ignore next */
    case '/': {
      return null
    }
    /* Recursive case. */
    default: {
      const elements = fs.readdirSync(dir)
      const includes = elements.map((name) => path.resolve(dir, name)).some(pattern)

      if (includes) return dir
      else return findUp(path.resolve(dir, '../'), pattern)
    }
  }
}
