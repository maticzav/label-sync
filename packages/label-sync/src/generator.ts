import type { LSCConfiguration, LSCRepository, LSCLabel } from '@labelsync/config'
import { Configurable } from './configurable'
import { Dict, withDefault, mapEntries } from './utils'
import { YAML } from './yaml'

/* Providers */

export function repo(repo: RepositoryInput): Repository {
  return new Repository(repo)
}

export function label(label: LabelInput | string, color?: string): Label {
  return new Label(label, color)
}

/* Classes */

/* Configuration */

export type ConfigurationInput = {
  repos: Dict<Repository>
}

export class Configuration extends YAML<LSCConfiguration> implements Configurable<LSCConfiguration> {
  private repositories: Dict<Repository> = {}

  constructor(config: ConfigurationInput) {
    super()
    this.repositories = config.repos
  }

  getConfiguration() {
    return {
      repos: mapEntries(this.repositories, (r) => r.getConfiguration()),
    }
  }
}

/* Repository */

export type RepositoryInput = {
  config?: RepositoryConfiguration
  labels: Label[]
}

export type RepositoryConfiguration = {
  removeUnconfiguredLabels?: boolean
}

/**
 * Repository configuration.
 */
export class Repository implements Configurable<LSCRepository> {
  private config: RepositoryConfiguration
  private labels: Label[] = []

  constructor(repo: RepositoryInput) {
    this.config = withDefault({}, repo.config)

    /* Copy over labels and skip duplicates. */
    for (const label of repo.labels.reverse()) {
      if (!this.labels.find((rl) => rl.getName() === label.getName())) {
        this.labels.push(label)
      }
    }
  }

  /**
   * Returns the collection of labels configured in repository.
   */
  *[Symbol.iterator]() {
    for (const label of this.labels) {
      yield label
    }
  }

  getConfiguration() {
    let labels: { [label: string]: LSCLabel } = {}

    /* Process labels */
    for (const label of this.labels) {
      const name = label.getName()
      /* istanbul ignore next */
      if (labels.hasOwnProperty(name)) {
        throw new Error(`Duplicate label ${name}`)
      }
      labels[name] = label.getConfiguration()
    }

    return {
      config: this.config,
      labels,
    }
  }
}

/* Label */

export type LabelInput =
  | {
      name: string
      color: string
      description?: string
      alias?: string[]
      siblings?: string[]
    }
  | string

export class Label implements Configurable<LSCLabel> {
  private name: string
  private color: string = ''
  private description: string | undefined
  private siblings: string[] = []
  private alias: string[] = []

  constructor(label: LabelInput | string, color?: string) {
    switch (typeof label) {
      case 'string': {
        this.name = label
        /* istanbul ignore next */
        if (!color) {
          throw new Error(`Label either accepts label(name, color) or label(config) object!`)
        }
        this.color = this.fixColor(color)
        return
      }
      case 'object': {
        this.name = label.name
        this.color = this.fixColor(label.color)
        this.description = label.description
        this.siblings = withDefault([], label.siblings)
        this.alias = withDefault([], label.alias)
        return
      }
    }
  }

  fixColor(color: string): string {
    if (!color.startsWith('#')) {
      return `#${color}`
    }
    return color
  }

  getName() {
    return this.name
  }

  getConfiguration() {
    return {
      color: this.color,
      description: this.description,
      alias: this.alias,
      siblings: this.siblings,
    }
  }
}
