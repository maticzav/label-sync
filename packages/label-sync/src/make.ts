import { writeFileSync } from 'fs'
import path from 'path'
import { isNull, isNullOrUndefined } from 'util'

import { findFileUp } from './fs'
import { Configuration } from './generator'
import { withDefault } from './utils'
import { EOL } from 'os'

/* Constants */

const LS_CONFIG_PATH = 'labelsync.yml'

/* File compilation */

export type MakeInput = {
  configs: Configuration[]
  outputs?: { config?: string }
}

/**
 * Parses a configuration file for the configuration.
 * @param param0 
 * @param cwd 
 */
export async function make(
  { configs, outputs }: MakeInput,
  cwd: string = process.cwd(),
): Promise<boolean> {
  /* Search for git folder */
  const pkgPath = await findFileUp(cwd, 'package.json')

  /* istanbul ignore next */
  if (isNull(pkgPath) && isNullOrUndefined(outputs?.config)) {
    return false
  }

  const output = withDefault(
    path.resolve(pkgPath!, LS_CONFIG_PATH),
    outputs?.config,
  )

  /* Write config to file */
  writeFileSync(output, configs.map(c => c.getYAML()).join(EOL), {
    encoding: 'utf-8',
  })

  return true
}
