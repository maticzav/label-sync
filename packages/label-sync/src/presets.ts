import { Label } from './generator'

export const colors = {
  neutral: '#EEEEEE',
  refine: '#fcaeec',
  shiny: '#3BDB8D',
  semiShiny: '#9cedc6',
  danger: '#FF5D5D',
  warning: '#FFCF2D',
  social: '#7057ff',
}

/**
 * Based off conventional commit notion of type. The kind of issue.
 * These labels may get their own colour to help visually differentiate
 * between them faster. The commit that closes this issue should generally
 * be of the same type as this label.
 */
export function type(name: string, color: string, description?: string): Label {
  return new Label({
    name: `type/${name}`,
    color: color,
    description: description,
  })
}

/**
 * Labels that help us track issue short-circuites or other minimal
 * categorical details.
 */
export function note(name: string, description?: string): Label {
  return new Label({
    name: `note/${name}`,
    color: colors.neutral,
    description: description,
  })
}

/**
 * Labels that help us track how impactful issues will be. Combined
 * with complexity label, helps inform prioritization.
 */
export function impact(name: string, description?: string): Label {
  return new Label({
    name: `impact/${name}`,
    color: colors.neutral,
    description: description,
  })
}

/**
 * Effort that help us track how impactful issues will be. Combined
 * with complexity label, helps inform prioritization.
 */
export function effort(name: string, description?: string): Label {
  return new Label({
    name: `effort/${name}`,
    color: colors.neutral,
    description: description,
  })
}

/**
 * Labels that help us mark issues as being on hold for some reason.
 */
export function needs(name: string, description?: string): Label {
  return new Label({
    name: `needs/${name}`,
    color: colors.warning,
    description: description,
  })
}

/**
 * Based off conventional commit notion of scope. What area of
 * the project does the issue touch. The commit that closes this issue
 * should generally be of the same scope as this label.
 */
export function scope(name: string, description?: string): Label {
  return new Label({
    name: `scope/${name}`,
    color: '#94ebfc',
    description: description,
  })
}

/**
 * Labels that help us coordinate with the community.
 */
export function community(name: string, description?: string): Label {
  return new Label({
    name: `community/${name}`,
    color: colors.social,
    description: description,
  })
}
