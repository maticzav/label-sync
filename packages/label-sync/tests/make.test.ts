import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import * as ls from '../src'
import { labelsync } from '../src'
import { parseConfig } from '../../../server/src/configuration'

const fsReadFile = promisify(fs.readFile)
const fsUnlink = promisify(fs.unlink)

describe('make:', () => {
  test('compiles configuration to path', async () => {
    const yamlPath = path.resolve(__dirname, 'labelsync.yml')

    labelsync(
      {
        repos: {
          'prisma-test-utils': ls.repo({
            config: {
              removeUnconfiguredLabels: true,
            },
            labels: [ls.label('bug/0-needs-reproduction', '#ff0022')],
          }),
        },
      },
      yamlPath,
    )

    const file = await fsReadFile(yamlPath, { encoding: 'utf-8' })
    await fsUnlink(yamlPath)

    expect(file).toMatchSnapshot()
  })

  test('integration test: compiles configuration to default path', async () => {
    const yamlPath = path.resolve(__dirname, './__fixtures__/labelsync.yml')

    labelsync(
      {
        repos: {
          'prisma-test-utils': ls.repo({
            config: {
              removeUnconfiguredLabels: true,
            },
            labels: [
              ls.label('kind/bug', '#02f5aa'),
              ls.label('bug/0-needs-reproduction', '#ff0022'),
              ls.label({
                name: 'bug/1-has-reproduction',
                color: '#ff0022',
                description: 'Indicates that an issue has reproduction',
                alias: ['bug'],
                siblings: ['kind/bug'],
              }),
              ls.type('bug', '#ff0022'),
              ls.note('sweet'),
              ls.impact('big'),
              ls.effort('low'),
              ls.needs('reproduction'),
              ls.scope('project'),
              ls.community('help-wanted'),
            ],
          }),
          'graphql-shield': ls.repo({
            config: {
              removeUnconfiguredLabels: true,
            },
            labels: [
              ls.label('kind/bug', '02f5aa'),
              ls.label('bug/0-needs-reproduction', '#ff0022'),
              ls.label({
                name: 'bug/1-has-reproduction',
                color: '#ff0022',
                description: 'Indicates that an issue has reproduction',
                alias: ['bug'],
                siblings: ['kind/bug'],
              }),
              ls.type('bug', '#ff0022'),
              ls.scope('project'),
              ls.community('help-wanted'),
            ],
          }),
        },
      },
      undefined,
      path.resolve(__dirname, './__fixtures__/'),
    )

    const file = await fsReadFile(yamlPath, { encoding: 'utf-8' })
    await fsUnlink(yamlPath)

    expect(file).toMatchSnapshot()

    const [errors, config] = parseConfig(
      {
        id: '1',
        createdAt: new Date(),
        updatedAt: new Date(),
        company: 'ACME',
        ghAccount: 'test',
        name: 'Foo Bar',
        email: 'foo@acme.com',
      },
      file,
    )
    expect(errors).toBeNull()
    expect(config).toMatchSnapshot()
  })
})
