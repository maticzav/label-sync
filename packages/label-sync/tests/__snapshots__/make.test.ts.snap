// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make: compiles configuration to path 1`] = `
"repos:
  prisma-test-utils:
    config:
      removeUnconfiguredLabels: true
    labels:
      bug/0-needs-reproduction:
        color: '#ff0022'
        alias: []
        siblings: []
"
`;

exports[`make: integration test: compiles configuration to default path 1`] = `
"repos:
  prisma-test-utils:
    config:
      removeUnconfiguredLabels: true
    labels:
      type/bug:
        color: '#ff0022'
        alias: []
        siblings: []
      bug/1-has-reproduction:
        color: '#ff0022'
        description: Indicates that an issue has reproduction
        alias:
          - bug
        siblings:
          - kind/bug
      kind/bug:
        color: '#02f5aa'
        alias: []
        siblings: []
  changed:
    config:
      removeUnconfiguredLabels: true
    labels:
      bug/1-has-reproduction:
        color: '#ff0022'
        description: Indicates that an issue has reproduction
        alias:
          - bug
        siblings:
          - kind/bug
      kind/bug:
        color: '#02f5aa'
        alias: []
        siblings: []
"
`;

exports[`make: integration test: compiles configuration to default path 2`] = `
Object {
  "repos": Object {
    "changed": Object {
      "config": Object {
        "removeUnconfiguredLabels": true,
      },
      "labels": Object {
        "bug/1-has-reproduction": Object {
          "alias": Array [
            "bug",
          ],
          "color": "ff0022",
          "description": "Indicates that an issue has reproduction",
          "siblings": Array [
            "kind/bug",
          ],
        },
        "kind/bug": Object {
          "alias": Array [],
          "color": "02f5aa",
          "siblings": Array [],
        },
      },
    },
    "prisma-test-utils": Object {
      "config": Object {
        "removeUnconfiguredLabels": true,
      },
      "labels": Object {
        "bug/1-has-reproduction": Object {
          "alias": Array [
            "bug",
          ],
          "color": "ff0022",
          "description": "Indicates that an issue has reproduction",
          "siblings": Array [
            "kind/bug",
          ],
        },
        "kind/bug": Object {
          "alias": Array [],
          "color": "02f5aa",
          "siblings": Array [],
        },
        "type/bug": Object {
          "alias": Array [],
          "color": "ff0022",
          "siblings": Array [],
        },
      },
    },
  },
}
`;
